src/Oscoin/Crypto/Address/Internal.hs

```diff
@@ -0,0 +1,112 @@
+--   |_  protocol version (bumped on hard forks)
+--
+-- * /23/ bytes for the CBOR-encoded address payload
+--
+-- * /6/ bytes for the checksum, computed from the <prefix + payload>.
+--
+data Address c = Address
+    { addressPrefix  :: AddressPrefix
+    , addressPayload :: AddressPayload c
+    } deriving (Eq, Show)
+
+data AddressPrefix = AddressPrefix
+    { protocolVersion :: ProtocolVersion
+    , addressType     :: AddressType
+    , addressFormat   :: AddressFormat
+    } deriving (Eq, Show)
+
+-- | A 'ProtocolVersion' is a byte representing the current version of the
+-- Oscoin protocol running. This makes possible to issue forks of the system
+-- by simply acting on this value, so that nodes running the old protocol
+-- won't be able to use addresses from the new fork. Conversely, nodes
+-- running the fork can be programmed to support both formats.
+newtype ProtocolVersion =
+    ProtocolVersion ProtocolVersionTag
+    deriving (Eq, Show)
+
+data ProtocolVersionTag =
+      ProtocolVersion_V1
+      deriving (Eq, Show)
+
+-- | The 'AddressType', i.e. the qualifying type of this 'Address'.
+-- Example:
+--
+-- 0x01 Mainnet
+-- 0x02 Testnet
+-- 0x03 Devnet
+-- ..
+newtype AddressType =
+    AddressType AddressTypeTag
+    deriving (Eq, Show)
+
+data AddressTypeTag =
+      AddressTypeMainnet -- (Mainnet, Production)
+    | AddressTypeTestNet -- (Testnet, Production)
+    | AddressTypeDevnet  -- (Devnet,  Production)
+    deriving (Eq, Show)
+
+-- | The 'AddressFormat' allows for flexibility in the 'Address' payload, i.e.
+-- it specifies the serialisation format used to encode the 'AddressPayload'.
+-- For now we have simply:
+--
+-- 0x01 -- CBOR
+--
+newtype AddressFormat =
+    AddressFormat AddressFormatTag
+    deriving (Eq, Show)
+
+data AddressFormatTag =
+    AddressFormatTagCBOR
+    deriving (Eq, Show)
+
+-- | The 'AddressPayload' is simply the hash of a 'PublicKey'.
+-- NOTE(adn): The use of 'Tagged' here is due to the fact we cannot encode
+-- directly 'ShortHashed c (PublicKey c)' as there is no isomorphism between
+-- a 'ShortHashed' and a 'ByteString', i.e. when decoding we wouldn't otherwise
+-- be able to go from a 'ByteString' to a 'ShortHashed' in a meaningful way.
+--
+newtype AddressPayload c =
+    AddressPayload { getAddressPayload :: Tagged (ShortHashed c (PublicKey c)) ByteArray }
+    deriving (Eq, Show)
+
+newtype Checksum = Checksum { fromChecksum :: ByteString }
+    deriving (Show, Eq)
+
+{------------------------------------------------------------------------------
+  Miscellanea functions
+------------------------------------------------------------------------------}
+
+configToAddressTypeTag :: Network -> Maybe AddressTypeTag
+configToAddressTypeTag = \case
+  Mainnet -> Just AddressTypeMainnet
+  Testnet -> Just AddressTypeTestNet
+  Devnet  -> Just AddressTypeDevnet
+  -- NOTE(adn) The 'Somenet' constructor is not exported, which makes this
+  -- pattern-match non-exhaustive.
```

Comment:
Somenet? :confused:

Reply:
Ok I just found out about `Somenet` ... do we want to export it and reserve `0x0` for it? Or perhaps the `Word8` should be gotten from an `Enum` instance on `Network`? @kim ?

Reply:
`Somenet` came from the strange situation that, when using multicast discovery, you want to make sure to not accidentally joining a network you didn’t mean to. So, it’s essentially “randomnet”. 

The same applies if we want to allow namespaced deployments to k8s - the DNS discovery would assume the namespace to equal the network name.

Idk, maybe there’s some other way to distinguish these.

Reply:
If we do export `Somenet` that would already be a step forward, because in this pattern match (which can now be made exhaustive) we can yield `Nothing`. This would affect only a situation where the user (or the node) purposefully call `fromPublicKey` with a somennet, which will be rejected anyway from the system.

Forward compatibility in not an issue -- when `Ubernet` will be added (say) we will simply associate a fresh `Word8` to it. I think that the morally correct thing is to _not_ assign `somenet` any word8. It's essentially "user error" and as such it shouldn't be handled/represented in the system.

Reply:
:+1: